### ISSUE TITLE
hadress_water/water.py example

### ISSUE THREAD

**Issue by almeida85** [espressopp/espressopp#309]
Hello,

I am playing with the `hadress_water/water.py` to learn AdresS. My end problem is based in a similar system of the example, but with a sphere configuration of the atomistic region. About this, I have several questions:

1 - I did a 100 ps test run in a workstation using 8 cores (as `mpirun -n 8 ./water.py`) and the run took 4 days (!). I want to test it in a cluster to see if I can get any acceleration because the production run is intended to be 1 ns. Are there any tips/tricks for running Espresso++ in a cluster? (The compilation itself has proven to be non-trivial.) I am interested in this issue after this post, https://github.com/espressopp/espressopp/issues/306#issuecomment-656346418, in my previous thread.

2 - As part of my problem I need to test several sizes of the atomistic region. The example provides the CG-CG interaction tables (obtained with IBI (VOTCA), I guess). If the size of the atomistic region is changed, the correspondent CG-CG interaction table has to be computed, right?

3 - The example provides a `grompp.mdp` file to show the GROMACS parser. Could I use this file, modified for AdresS to run the same simulation using GROMACS (I am more familiar with it)? I want to compare the performance between Espresso++ and GROMACS 5.1.5, in a cluster, using this system.

4 - How to apply the Hamiltonian pressure and density compensation to the code? (In order to reproduce the results of DOI:10.1140/epjst/e2016-60151-6).

Thanks a lot in advance for your help and suggestions.

Best,
Yasser

**Comment by junghans** [comment]
> 2 - As part of my problem I need to test several sizes of the atomistic region. The example provides the CG-CG interaction tables (obtained with IBI (VOTCA), I guess). If the size of the atomistic region is changed, the correspondent CG-CG interaction table has to be computed, right?

Usually the CG-CG interactions are transferable between different sizes of atomistic regions.

> 3 - The example provides a `grompp.mdp` file to show the GROMACS parser. Could I use this file, modified for AdresS to run the same simulation using GROMACS (I am more familiar with it)? I want to compare the performance between Espresso++ and GROMACS 5.1.5, in a cluster, using this system.

I think that was the idea, but it won't hurt to consult the 5.1.5 gromacs manual as well.

**Comment by jkrajniak** [comment]
> 1 - I did a 100 ps test run in a workstation using 8 cores (as `mpirun -n 8 ./water.py`) and the run took 4 days (!). I want to test it in a cluster to see if I can get any acceleration because the production run is intended to be 1 ns. Are there any tips/tricks for running Espresso++ in a cluster? (The compilation itself has proven to be non-trivial.) I am interested in this issue after this post, [#306 (comment)](https://github.com/espressopp/espressopp/issues/306#issuecomment-656346418), in my previous thread.
> 
Try to compile E++ with appropriate flags for the cluster architecture (maybe obvious, but could speed up)

> 4 - I modified the script to compute the density and pressure profiles, as in `hadress_tetraliquid_plain/hadress.py`. How to compute these profiles post-run?
> 
The density can be easily computed from the trajectory. As for the pressure profiles, I suppose you need to store forces. This can
be e.g. handled by H5MD trajectory writer.

> 5 - How to apply the Hamiltonian pressure and density compensation to the code? (In order to reproduce the results of DOI:10.1140/epjst/e2016-60151-6, Fig. 5).
> 
The best is to contact authors of the paper. As I see they used LAMMPS with their own package. Maybe there is the code.

**Comment by junghans** [comment]
> > 5 - How to apply the Hamiltonian pressure and density compensation to the code? (In order to reproduce the results of DOI:[10.1140/epjst/e2016-60151-6](https://doi.org/10.1140/epjst/e2016-60151-6), Fig. 5).
> 
> The best is to contact authors of the paper. As I see they used LAMMPS with their own package. Maybe there is the code.

Oh right, that is implemented in https://github.com/hadress/lammps/tree/USER-HADRESS

**Comment by almeida85** [comment]
And what about `hadress_tetraliquid/hadress_tetraliquid_FEC`? The `DensityProfile_Gibbs_Reference.dat` file in the example shows a profile similar to the goal behavior. Is the density correction implemented there? I am diving into the code.

**Comment by almeida85** [comment]
In the `espressopp.integrator.FreeEnergyCompensation` page of the User interface, it states: 
"Free Energy Compensation used in Hamiltonian Adaptive Resolution Simulations (H-AdResS) or Path Integral Adaptive Resolution Simulations (PI-AdResS). This works for spherical or slab adaptive resolution geometries. **However, it only works for fixed, non-moving atomistic region (otherwise, H-AdResS is not properly defined).**"

What "fixed, non-moving atomistic region means"? Does it mean that the atomistic region doesn't exchange with the CG region on the fly?

**Comment by govarguz** [comment]
It means that the atomistic region _"itself"_ is geometrically defined in the simulation box as fixed (e.g. for a slab in the x-axis x_0=0.0 to x_{at}=20.0, where x_{total}=40). However, the on-the-fly exchange of particles between resolutions is not affected.

**Comment by jkrajniak** [comment]
> In the `espressopp.integrator.FreeEnergyCompensation` page of the User interface, it states:
> "Free Energy Compensation used in Hamiltonian Adaptive Resolution Simulations (H-AdResS) or Path Integral Adaptive Resolution Simulations (PI-AdResS). This works for spherical or slab adaptive resolution geometries. **However, it only works for fixed, non-moving atomistic region (otherwise, H-AdResS is not properly defined).**"
> 
> What "fixed, non-moving atomistic region means"? Does it mean that the atomistic region doesn't exchange with the CG region on the fly?

No, it does not have anything to do with exchanging particles between regions. In case of F-AdResS it is possible to define the atomistic region with respect to a particle. Effectively the atomistic region "moves" inside the simulation box.

Check `src/VerletListAdress.py`, there is an example of how it works:
`espressopp.VerletListAdress(system, cutoff=rc, adrcut=rc, dEx=ex_size, dHy=hy_size, pids=[adrCenterPID], sphereAdr=True)`

**Comment by almeida85** [comment]
> It means that the atomistic region _"itself"_ is geometrically defined in the simulation box as fixed (e.g. for a slab in the x-axis x_0=0.0 to x_{at}=20.0, where x_{total}=40). However, the on-the-fly exchange of particles between resolutions is not affected.

Gotcha.

So, following up with my previous question, can I use the `table_FEC_Gibbs.dat` and the `espressopp.integrator.FreeEnergyCompensation` as in `hadress_tetraliquid_FEC/hadressDensityFEC.py` to do the density correction for `water.py`?

**Comment by almeida85** [comment]
> Try to compile E++ with appropriate flags for the cluster architecture (maybe obvious, but could speed up)

Which are the appropriate flags for cluster compilation? (apart from MPI and OpenMPI flags). I am looking in the `CMakeCache.txt` file.

**Comment by jkrajniak** [comment]
> > It means that the atomistic region _"itself"_ is geometrically defined in the simulation box as fixed (e.g. for a slab in the x-axis x_0=0.0 to x_{at}=20.0, where x_{total}=40). However, the on-the-fly exchange of particles between resolutions is not affected.
> 
> Gotcha.
> 
> So, following up with my previous question, can I use the `table_FEC_Gibbs.dat` and the `espressopp.integrator.FreeEnergyCompensation` as in `hadress_tetraliquid_FEC/hadressDensityFEC.py` to do the density correction for `water.py`?

Well, you have to check what does it mean to your system if you use the values from that table. How that table was created and so on. It is rather a scientific/project question not really related to E++ project itself.

**Comment by jkrajniak** [comment]
> > Try to compile E++ with appropriate flags for the cluster architecture (maybe obvious, but could speed up)
> 
> Which are the appropriate flags for cluster compilation? (apart from MPI and OpenMPI flags). I am looking in the `CMakeCache.txt` file.

This is an environment dependent question. You should set your cmake build type to `Release` which implies `-O3` flags. Then, if your system has toolchain `intel` then you should definitely use `icc` compiler. If so, you can use `-xHost` flag to optimize the compilation for the platform. But, keep in mind that if you are compiling in so called login-nodes then the architecture will be selected from login node. For best, ask your local HPC support team how to compile C++ code in most optimal way.

**Comment by almeida85** [comment]
> It is rather a scientific/project question not really related to E++ project itself.

In a way, I think could be related to the E++ project because it would be nice if the water example also includes the density and pressure correction to compare with the non-corrected (current) system. In the end, water systems are the most popular.
